gateway/mw_streaming.go

```diff
@@ -0,0 +1,383 @@
+	sm.mw.Logger().Debugf("Creating stream: %s", streamFullID)
+
+	stream := streaming.NewStream(sm.mw.allowedUnsafe)
+	err := stream.Start(config, &handleFuncAdapter{
+		mw:       sm.mw,
+		streamID: streamFullID,
+		muxer:    sm.muxer,
+		sm:       sm,
+		// child logger is necessary to prevent race condition
+		logger: sm.mw.Logger().WithField("stream", streamFullID),
+	})
+	if err != nil {
+		sm.mw.Logger().Errorf("Failed to start stream %s: %v", streamFullID, err)
+		return err
+	}
+
+	sm.streams.Store(streamFullID, stream)
+	sm.mw.Logger().Infof("Successfully created stream: %s", streamFullID)
+
+	return nil
+}
+
+func (sm *StreamManager) hasPath(path string) bool {
+	for _, p := range sm.listenPaths {
+		if strings.TrimPrefix(path, "/") == strings.TrimPrefix(p, "/") {
+			return true
+		}
+	}
+	return false
+}
+
+// ProcessRequest will handle the streaming functionality
+func (s *StreamingMiddleware) ProcessRequest(w http.ResponseWriter, r *http.Request, _ interface{}) (error, int) {
+	strippedPath := s.Spec.StripListenPath(r.URL.Path)
+	if !s.defaultStreamManager.hasPath(strippedPath) {
+		return nil, http.StatusOK
+	}
+
+	s.Logger().Debugf("Processing request: %s, %s", r.URL.Path, strippedPath)
+
+	newRequest := r.Clone(r.Context())
+	newRequest.URL.Path = strippedPath
+
+	var match mux.RouteMatch
+	streamManager := s.createStreamManager(r)
+	streamManager.routeLock.Lock()
+	streamManager.muxer.Match(newRequest, &match)
+	streamManager.routeLock.Unlock()
+
+	// direct Bento handler
+	handler, _ := match.Handler.(http.HandlerFunc)
+
+	handler.ServeHTTP(w, r)
+
+	return nil, mwStatusRespond
+}
+
+func (s *StreamingMiddleware) Unload() {
+	s.Logger().Debugf("Unloading streaming middleware %s", s.Spec.Name)
+
+	totalStreams := 0
+	s.streamManagers.Range(func(_, value interface{}) bool {
+		manager := value.(*StreamManager)
+		manager.streams.Range(func(_, _ interface{}) bool {
+			totalStreams++
+			return true
+		})
+		return true
+	})
+	globalStreamCounter.Add(-int64(totalStreams))
+
+	s.cancel()
+
+	s.Logger().Debug("Closing active streams")
+	s.streamManagers.Range(func(_, value interface{}) bool {
+		manager := value.(*StreamManager)
+		manager.streams.Range(func(_, streamValue interface{}) bool {
+			if stream, ok := streamValue.(*streaming.Stream); ok {
+				stream.Reset()
+			}
+			return true
+		})
+		return true
+	})
+
+	s.streamManagers = sync.Map{}
+
+	s.Logger().Info("All streams successfully removed")
+}
+
+type handleFuncAdapter struct {
+	streamID string
+	sm       *StreamManager
+	mw       *StreamingMiddleware
+	muxer    *mux.Router
+	logger   *logrus.Entry
+}
+
+func (h *handleFuncAdapter) HandleFunc(path string, f func(http.ResponseWriter, *http.Request)) {
+	h.logger.Debugf("Registering streaming handleFunc for path: %s", path)
+
+	if h.mw == nil || h.muxer == nil {
+		h.logger.Error("StreamingMiddleware or muxer is nil")
+		return
+	}
+
+	h.sm.routeLock.Lock()
```

Comment:
I don't fully understand why we need this mutex. It protects the muxer. The other use is in `handleFuncAdapter.HandleFunc` but this method should be called before calling the `ProcessRequest`. 

```go
streamManager.routeLock.Lock()
streamManager.muxer.Match(newRequest, &match)
streamManager.routeLock.Unlock()
```

I think `HandleFunc` must be called before `ProcessRequest` calls `muxer.Match` because the stream has created and started in `ProcessRequest` method. If not, there is no guarantee that the endpoint has registered before calling `muxer.Match`. 

If there is an uncertainty in establishing *happened before* relationship between processing the request and registering the handler, we might consider adding an health check method to `Stream.Start` method. Currently, it only waits for 100 milliseconds before returning.

Reply:
Yeah the lack of a mutex triggered an error when testing for race condition, because the router is modified by registering routes via benthos and accessed via in the handler, and it is read by ProcessRequest on a different goroutine.